### ISSUE TITLE
Implement @embed as a replacement for bin2v

### ISSUE THREAD

**Issue by radare** [vlang/v#5169]
I propose to implement this feature which is available in other languages like Zig

* https://ziglang.org/documentation/master/#toc-embedFile

The idea behind this compile-time keyword is to replace the statement for a byte array like bin2v do:

```go
$ v bin2v /bin/sleep |head
module main
const (
  sleep_len
  sleep = [ byte(207),...]
```

The actual problem in bin2v is that this array of bytes is not really `const`, by looking at the generated C code i would expect to see something like:

```c
const unsigned char foo[] = { 1,2,3, 4, 5};

main() {
	printf ("%d\n", sizeof(foo)); // 5 == array length (foo.len from V)
	printf ("%d\n", foo[0]); // returns 1
	// foo[0] = 1;
}
```

What V is actually generating is a global varable that is  initialized in the heap with an extra memcpy making the bootstrap slower than it should and not making it really const.

```c
        _const_foo = new_array_from_c_array(4, 4, sizeof(byte), _MOV((byte[4]){
((byte)(1)), 2, 3, 4,
}));

static array  new_array_from_c_array(int len, int cap, int elm_size, voidptr c_array) {
        int cap_ = (cap < len ? (
                len
        ) : (
                cap
        ));
        array arr = (array){
                .element_size = elm_size,
                .data = vcalloc(cap_ * elm_size),
                .len = len,
                .cap = cap_,
        };
        memcpy(arr.data, c_array, len * elm_size);
        return arr;
}
```

To summarize, this is my proposal:

```go
module main

const (
  image_jpg := @embed('file.jpg')
)

fn main() {
  println('embedded image size: ${image_jpg.len}')
}
```

This statement should only work for global variables (const or not)

**Comment by spytheman** [comment]
https://github.com/vlang/rfcs/blob/master/embedding_resources.md

**Comment by JalonSolov** [comment]
Just to have it all in one place, Alex expressed a preference for `$embed` rather than `@embed`.

https://discord.com/channels/592103645835821068/592114487759470596/717367581215948861

I don't know if I am correct, but this would seem more in line with the current $if, etc. which control code generation, as opposed to `@VEXE` or `@VLIB` which are basically just paths.

**Comment by medvednikov** [comment]
$embed_file actually, to be more explicit

**Comment by trufae** [comment]
So the final decision is to use:

```go
module main

const (
  image_jpg := $embed_file('file.jpg')
)

fn main() {
  println('embedded image size: ${image_jpg.len}')
}
```

**Comment by medvednikov** [comment]
correct

**Comment by trufae** [comment]
What about $embed.file instead of $embed_file. The only comptimecall i see in V is $vweb and it's only used for $vweb.html()

**Comment by medvednikov** [comment]
no need

there's also $tmpl('template.txt') now, not pushed yet

**Comment by Larpon** [comment]
Just a, more or less, relevant note:
We also have `@FN`, `@MOD`, `@STRUCT` which will result in direct string replacement at compile time.
:+1: for `$embed.file()`

**Comment by Dialga** [comment]
This has now been implemented by #8048, in the form of `$embed_file()`.

**Comment by Larpon** [comment]
Yes this is now working 🙂

**Comment by trufae** [comment]
Awesome!!! I’m gonna try it right now!
